tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,146 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Discriminative Layer Training Optimizer for TensorFlow."""
+
+import tensorflow as tf
+from typeguard import typechecked
+from typing import Union
+from tensorflow.keras.optimizers import Optimizer
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class MultiOptimzer(Optimizer):
+    @typechecked
+    def __init__(
+        self,
+        optimizer_layer_pairs: Union[list, None] = None,
+        optimizer_specs: Union[list, None] = None,
+        name: str = "MultiOptimzer",
+        **kwargs
+    ):
+
+        """
+        Creates a wrapper around a set of instantiated optimizer layer pairs. Allows for implementation of
+        Discriminative Layer Training as per https://arxiv.org/abs/1607.05440. Generally useful for transfer learning
+        of deep networks.
+
+        Each optimizer will optimize only the weights associated with its paired layer. This can be used
+        to implement discriminative layer training by assigning different learning rates to each optimizer
+        layer pair. (Optimizer, list(Layers)) pairs are also supported. Please note that the layers must be
+        instantiated before instantiating the optimizer.
+
+        Currently, MultiOpt does not support callbacks that modify optimizers. However, you can instantiate
+        optimizer layer pairs with tf.keras.optimizers.schedules.LearningRateSchedule instead of a static learning
+        rate.
+
+        This code should function on CPU, GPU, and TPU. Apply the with strategy.scope() context as you
+        would with any other optimizer.
+
+        Example:
+
+        ```python
+
+        model = get_model()
+
+        opt1 = tf.keras.optimizers.Adam(learning_rate=1e-4)
+        opt2 = tf.keras.optimizers.Adam(learning_rate=1e-2)
+
+        opt_layer_pairs = [(opt1, model.layers[0]),
+                            (opt2, model.layers[1:])]
+
+        loss = tf.keras.losses.MSE
+        optimizer = tfa.optimizers.MultiOpt(opt_layer_pairs)
+
+        model.compile(optimizer=optimizer, loss = loss)
+
+        model.fit(x,y)
+
+        ```
+
+
+        """
```

Comment:
This should be put between `class MultiOptimizer` and `def __init__`. We also miss a section of `Args`. It's also encouraged to use doctest for python code snippet (optional, if you have time)

See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/optimizer_v2/adam.py#L47-L69